infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvironmentVariablesSorter.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.provision.env;
+
+import static java.lang.String.format;
+import static java.util.Comparator.naturalOrder;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.fabric8.kubernetes.api.model.EnvVar;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+
+/**
+ * If there are any references to other environment variables in the value of some environment
+ * variable, the "dependent" environment variable needs to be defined "later" than the env var it
+ * depends on for Kubernetes variable expansion to work.
+ *
+ * <p>This class makes sure that the environment variables are sorted.
+ */
+final class EnvironmentVariablesSorter implements Comparator<EnvVar> {
```

Comment:
This class IS a comparator and it is USED as one. It is used by the single static method, called `sort` that happens to be located within the class itself. That doesn't make the class itself use a comparator. 

Static methods, if not referencing to some static state, which is evil in its own regard, are themselves stateless and essentially independent of the class their placed in. Placing a static method into a class is basically namespacing, nothing more.

Reply:
I would expect next usage from a class that is comparator like it is now:
```
List<EnvVar> ret = new ArrayList<>(envVars.values());
ret.sort(sorter);
```
where construction of the object and passing the same data to its constructor seems unneccessary.
Because of that, it seems to me that this util class should not serve as both comparator and util class that uses the comparator.

Reply:
I also think that `inheritance` is not right thing there.
I see several possible ways here: Have `EnvironmentVariablesSorter` with `EnvVarComparator implements Comparator<EnvVar>` as an internal class. Then `List<EnvVar> sort(Collections EnvVar> envVars)` may be not static method, and tests would looks better. Like `EnvVarsConverterTest` would just test `EnvVarsConverter` functionality (override environment with machine provided, putting sorted value (but sorting is mocked)) and `EnvironmentVariablesSorter` would test different sort cases - like sort, do nothing if not needed, cyclic dependencies detection.

Or have `EnvVarComparator` only and then EnvVarsConverter may use it. The following example looks OK for me but then `EnvVarsConverterTest` would test two separate classes
```
        // the env vars defined in our machine config take precedence over the ones already defined
        // in the container, if any
        container.getEnv().forEach(v -> envVars.putIfAbsent(v.getName(), v));
        container.setEnv(new ArrayList<>(envVars.values()));

        // The environment variable interpolation only works if a variable that is referenced
        // ....
        container.getEnv().sort(new EnvVarComparator(container.getEnv()));
```